src/consensus/upgrades.cpp

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2018 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "consensus/upgrades.h"
+
+const struct NUInfo NetworkUpgradeInfo[Consensus::MAX_NETWORK_UPGRADES] = {
+    {
+        /*.nBranchId =*/ 0x00000000,
+        /*.strName =*/ "Test dummy",
+        /*.strInfo =*/ "Test dummy info",
+    }
+};
+
+UpgradeState NetworkUpgradeState(
```

Comment:
Just to see I understand:  there is no way to make an Upgrade for a limited range of blocks.  Once it's active, it is active for ever - unless you change it's activation height to `NO_ACTIVATION_HEIGHT` which would require a hard fork

Reply:
There's a confusion between upgrades and epochs here, which maybe needs clarification in the ZIP:

- An upgrade is a single event, with blocks before and after it. It doesn't have a duration.
- An epoch is the period between two consecutive upgrades. It does have a duration, either explicit (between two upgrades) or implicit (from last upgrade until now).

So if you want some rule to be only applied to a specific epoch, you could use `CurrentEpoch()` API.

Reply:
The ZIP now defines a "branch" vs a "hard fork", which captures the above distinction (epochs are sub-regions of branches, and upgrades are subsets of hard forks).

Reply:
OK, so if you want a change to be for a limited range of blocks, you have to define in advance two network upgrades that are the endpoints of this range.